pkg/debuginfo/store.go

```diff
@@ -192,87 +205,118 @@ func (s *Store) Upload(stream debuginfopb.DebugInfoService_UploadServer) error {
+		}
+
 		objFile, err := s.fetchObjectFile(ctx, buildID)
 		if err != nil {
 			return status.Error(codes.Internal, err.Error())
 		}
 
-		if req.GetInfo().Hash != "" {
-			h, err := hash.File(objFile)
-			if err != nil {
-				return status.Error(codes.Internal, err.Error())
-			}
-
-			if h == req.GetInfo().Hash {
-				level.Debug(s.logger).Log("msg", "debug info already exists", "buildid", buildID)
-				return status.Error(codes.AlreadyExists, "debuginfo already exists")
+		if err := elfutils.ValidateFile(objFile); err != nil {
+			// Failed to validate. Mark the file as corrupted, and let the client try to upload it again.
+			if err := s.metadataManager.corrupted(ctx, buildID); err != nil {
+				level.Warn(s.logger).Log("msg", "failed to update metadata as corrupted", "err", err)
 			}
+			// Client will retry.
+			return status.Error(codes.Internal, err.Error())
 		}
 
+		// Valid.
 		hasDWARF, err := elfutils.HasDWARF(objFile)
-		if err != nil && !errors.Is(err, io.EOF) {
-			var fe *elf.FormatError
-			if errors.As(err, &fe) {
-				// Ignore bad magic number if all zero
-				if !strings.Contains(fe.Error(), "bad magic number '[0 0 0 0]'") {
-					return status.Error(codes.Internal, err.Error())
-				}
-			} else {
-				return status.Error(codes.Internal, err.Error())
-			}
+		if err != nil {
+			level.Debug(s.logger).Log("msg", "failed to check for DWARF", "err", err)
 		}
 		if hasDWARF {
-			// We probably have the best version.
-			level.Debug(s.logger).Log("msg", "debug info with DWARF already exists", "buildid", buildID)
 			return status.Error(codes.AlreadyExists, "debuginfo already exists")
 		}
 	}
 
-	if err := s.metadataManager.update(ctx, buildID, metadataStateUploading); err != nil {
-		level.Error(s.logger).Log("msg", "failed to update metadata", "err", err)
+	if err := s.metadataManager.uploading(ctx, buildID); err != nil {
+		err := fmt.Errorf("failed to update metadata before uploading: %w", err)
+		return status.Error(codes.Internal, err.Error())
 	}
 
-	// At this point we know that we still have a better version of the debug information file,
+	//nolint:forcetypeassert
+	b := s.pool.Get().([]byte)
+	buf := flexbuf.With(b)
+	defer func() {
+		// TODO(kakkoyun): Check if this creates any problems with reused buffers.
+		buf.Release()
+		b = b[:0]
+	}()
+
+	// At this point we know that we received a better version of the debug information file,
 	// so let the client upload it.
 	r := &UploadReader{stream: stream}
-	if err := s.bucket.Upload(ctx, objectPath(buildID), r); err != nil {
+	if err := s.bucket.Upload(ctx, objectPath(buildID), io.TeeReader(r, buf)); err != nil {
 		msg := "failed to upload"
 		level.Error(s.logger).Log("msg", msg, "err", err)
 		return status.Errorf(codes.Unknown, msg)
 	}
 
-	if err := s.metadataManager.update(ctx, buildID, metadataStateUploaded); err != nil {
-		level.Error(s.logger).Log("msg", "failed to update metadata", "err", err)
+	if err := elfutils.ValidateReader(buf); err != nil {
```

Comment:
I don't think there's a better way to do this, but I am wondering what can we do to reduce memory usage.

My concern is that we will store the debug info files we are uploading in memory which can put a lot of memory pressure in the box. By using the pool, we will make sure that we don't allocate too much, which is good! But the memory will still be used, and when several multi-gigabyte debug files are uploaded at the same time, we will have to store them in memory.

Please correct me if I am wrong and happy to hear what you think!

Reply:
You're definitely right. That's why this is the last piece that I have added. I really don't want to do this, but it's the only way to validate completely what we have received. If we're willing to loosen up the guarantees, we can get by just reading the first few bytes. And if something happens during symbolization, we can mark this file as corrupted (which we already do). We just need to make sure the error that we receive is not recoverable. (And this is what I meant by the last to-do item).